test/test_model.py

```diff
@@ -758,6 +843,52 @@ def test_storage_tool_errors(self):
                 run()
             self.assertEqual(fake_script_calls(self, clear=True), calls)
 
+    def test_function_tool_errors(self):
+        test_cases = [(
+            lambda: fake_script(self, 'function-get', f'echo fooerror >&2 ; exit 1'),
+            lambda: self.backend.function_get(),
+            ops.model.ModelError,
+            [['function-get', '--format=json']],
+        ), (
+            lambda: fake_script(self, 'function-set', f'echo fooerror >&2 ; exit 1'),
+            lambda: self.backend.function_set(foo='bar', dead='beef'),
+            ops.model.ModelError,
+            [["function-set", "foo='bar'", "dead='beef'"]],
+        ), (
+            lambda: fake_script(self, 'function-fail', f'echo fooerror >&2 ; exit 1'),
+            lambda: self.backend.function_fail('fail-message'),
+            ops.model.ModelError,
+            [["function-fail", "'fail-message'"]],
+        )]
+        for do_fake, run, exception, calls in test_cases:
+            do_fake()
+            with self.assertRaises(exception):
+                run()
+            self.assertEqual(fake_script_calls(self, clear=True), calls)
```

Comment:
Given this loop, which consists almost exclusively of calling the individual lambdas right above, it looks like the "table" style is damaging readability here instead of helping. Also, this is leaving the prior faked scripts behind, which is worse than not doing that.

I suspect having actual independent test cases here will be both cleaner and more readable.